pkg/eventqueue/eventqueue.go

```diff
@@ -169,11 +175,15 @@ func (ev *Event) WasCancelled() bool {
 // closed, indicating that the Event was not ran. This function may block if
 // the queue is at its capacity for events.
 func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
-
 	if q.notSafeToAccess() || ev == nil {
 		return nil
 	}
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
```

Comment:
ideally this should be a sync.Once, otherwise the event order is not being kept, i.e. if we enqueue the same event at the same time, the response of the 2nd queued might, and probably will, be returned before the 1st queued is executed.

Reply:
If we Enqueue the same event at the same time, one of the events will add to the int atomically, and the other one won't - how does that differ from a `sync.Once`? I can use a `sync.Once` since that is basically what I am using here (`sync.Once` is a wrapper around a mutex and `sync.Atomic` loading and storing a `uint32`. 

In other words, ordering will never be guaranteed in this case, it's a matter of who gets there first. I think that's fine. The `EventQueue` wasn't designed to have the same `Event` enqueued multiple times; safeguarding against that with some state which signifies whether it has been Enqueued is a needed fix, but I do not think that we need to ensure that if the same event is enqueued multiple times that we need to preserve the ordering of who calls Enqueue - that would mean that multiple goroutines would be Enqueueing the same event at the same time.

Reply:
>how does that differ from a sync.Once?

The sync.Once will wait until all other go routines being ran inside the Once are executed.

Reply:
I do not understand what advantages of waiting for the goroutines being ran inside the `sync.Once` are executed provides us is in this case. The critical section here is quite small as opposed to wrapping the contents of `Enqueue` itself within a `sync.Once`. Can you elaborate on what advantages that provides in this case?

Reply:
>Can you elaborate on what advantages that provides in this case?

Why is the function named "Enqueue" if we will return earlier in a 2nd, and following "enqueues" of an event that is still in the queue and still being processed? By earlier I meant that we will return even before the event was processed the first time.

Reply:
The first enqueue will process the event, the second enqueue will return a nil result immediately, which the user can check for. A `nil` return means that the Enqueue was a no-op. I can write that into the documentation. When would the ordering of returns matter in this case? I do not ever see a case in which a user would enqueue the same event twice across different goroutines. With this PR, we guard against that by returning a nil value for the second call to Enqueue. Can you provide an example where the ordering would matter? I feel like this is a semantic argument as opposed to one of correctness.

Reply:
It's weird to have a function called enqueue for which in reality is not enqueing events.

Reply:
The code was designed such that a given event should be enqueued only once. If an event is enqueued more than once, we return a signal to the user (`nil` result) that it was not enqueued. I could return an error as well from `Enqueue` indicating as such? If a user wants to Enqueue similar information again, they can call `NewEvent` and Enqueue that information. The path you are talking about is not a use-case that was envisioned when I wrote the code, nor is it how it is being used throughout Cilium.

Reply:
To summarize: the `nil` return is an error condition the caller should check for.